Documentation: Limitations (#714)
model-checking/rmc: rmc-docs/src/limitations.md

| | Dereferencing unaligned raw pointers | No | |
| | Breaking pointer aliasing rules | No | |
| | Mutating immutable data | No | |
| | Invoking undefined behavior via compiler intrinsics | Partial | See [Observations - Intrinsics](#intrinsics) |
| | Executing code compiled with platform features that the current platform does not support | No | |
| | Producing an invalid value, even in private fields and locals | No | |

## Observations on partially or unsupported features

### Code generation for unsupported features

RMC aims to be a industrial verification tool. Most industrial crates may
include unsupported features in parts of their code that do not need to be
verified. In general, this should not prevent users from verifying their code.

Because of that, the general rule is that RMC generates an `assert(false)`
statement when compiling any unsupported feature. This will cause proofs to fail
if the statement is reachable during the verification stage. However, the
analysis will not affected if the statement is not reachable from the code under
verification, so users can still verify components of their code not using
unsupported features.

### Assembly

> CBMC support for assembly code is quite limited. It may be extended in the

Reviewer: I would say RMC instead of CBMC here. At this point, I would only mention CBMC if you want to add some link to CBMC assembly support.

future but at present there are no plans to do so.

Check out the tracking issues for [inline assembly (`asm!`
macro)](https://github.com/model-checking/rmc/issues/2) and [global assembly
(`asm_global!` macro)](https://github.com/model-checking/rmc/issues/316) to know
more about the current status.

### Concurrency

Concurrent features are out of scope for RMC. In general, the verification of
concurrent programs continues to be an open research problem. In particular,

> support for concurrent verification in CBMC is limited and RMC emits a warning

Reviewer: The same here.

whenever it encounters concurrent code.

> ### Not inlined standard library functions

Reviewer: Maybe change the title to standard library functions.

At present, RMC is able to link in functions from the standard library but the

> generated code will not contain them unless they are inlined. This results in

Reviewer: We should add that we support to generic functions and macros at this point. They get monomorphized / expanded during the compilation. BTW, I don't think we are running the rustc pass to inline functions.

Author: Not sure why you want to add it here. I think the table already includes this information and I'm not sure what it has to do with the standard library.

Removed "unless they are inlined" from this block.

Reviewer: Sorry if I wasn't clear. I was talking about the "unless they are inlined". The correct would be "unless they are generic, intrinsics, inlined or macros." :)

verification failures if the code under verification, for example, includes a
`println!` statement.

> We have done some experiments to embed the standard library into the generated

Reviewer: Did we run these experiments before or after enabling the code slicing?

Author: If you refer to `--drop-unused-functions`, I tried both when doing these experiments. It reduced the generated code considerably, but still kept many other functions around which made this approach unfeasible.

Reviewer: did your experiments include tests with `println`?

Author: Yes, in fact they just included `println`.

code, but this causes verification times to increase significantly. As of now,
we have not been able to find a simple solution for [this
issue](https://github.com/model-checking/rmc/issues/581), but we have some ideas
for future work in this direction.

### Advanced features

> Advanced features (traits, types, etc.) from Rust have proven to particularly

Reviewer: I would say something like: the semantics around some of these rust advanced features are not formally defined which makes it harder to ensure that we can properly model all their use cases. If you discover any case that cannot be handled properly by RMC, please file a bug report.

Author: Thanks! I liked the first part a lot so I included it in here. Let me know if you want to change it back.

# Limitations

## Rust feature support

The table below tries to summarize the current support in RMC for
the Rust language features according to the [Rust Reference](https://doc.rust-lang.org/stable/reference/).
We use the following values to indicate the level of support:
 * **Yes**: The feature is fully supported. We are not aware of any issue with it.
 * **Partial**: The feature is at least partially supported. We are aware of some issue with
 with it.
 * **No**: The feature is not supported. Some support may be available but analyses should not be trusted.

As with all software, bugs may be found anywhere regardless of the level of support. In such cases, we
would greatly appreciate that you [filed a bug report](https://github.com/model-checking/rmc/issues/new?assignees=&labels=bug&template=bug_report.md).

> Reference | Feature | Support | Observations |

Reviewer 2: Would "Remarks" or "Notes" be more suitable than "Observations"?

10.9 | Lifetime elision | Yes | |
11 | Special types and traits | Partial | |
| | `Box<T>` | Yes | |
| | `Rc<T>` | Yes | |
| | `Arc<T>` | Yes | |
| | `Pin<T>` | Yes | |
| | `UnsafeCell<T>` | Partial | |
| | `PhantomData<T>` | Partial | *Review this* |
| | Operator Traits | Partial | |
| | `Deref` and `DerefMut` | Yes | |
| | `Drop` | Partial | |
| | `Copy` | Yes | |
| | `Clone` | Yes | |
14 | Linkage | Yes | |
15.1 | Unsafe functions | Yes | |
15.2 | Unsafe blocks | Yes | |
15.3 | Behavior considered undefined | Partial | |
| | Data races | No | See [Observations - Concurrency](#concurrency) |
| | Dereferencing dangling raw pointers | Yes | |
| | Dereferencing unaligned raw pointers | No | |
| | Breaking pointer aliasing rules | No | |
| | Mutating immutable data | No | |
| | Invoking undefined behavior via compiler intrinsics | Partial | See [Observations - Intrinsics](#intrinsics) |
| | Executing code compiled with platform features that the current platform does not support | No | |
| | Producing an invalid value, even in private fields and locals | No | |

## Observations on partially or unsupported features

### Code generation for unsupported features

> RMC aims to be a industrial verification tool. Most industrial crates may

Reviewer 2: ```suggestion
RMC aims to be an industrial verification tool. Most industrial crates may
```
Not sure I like the word "industrial", but can't think of an alternative.

Reviewer 2: One possibility (I don't feel strongly about it either): RMC is intended to be used on real-world Rust code. Most real-world crates ...

include unsupported features in parts of their code that do not need to be

> verified. In general, this should not prevent users from verifying their code.

Reviewer 2: ```suggestion
verified. In general, this should not prevent users from using RMC to verify their code.
```

> Because of that, the general rule is that RMC generates an `assert(false)`

Reviewer 2: Technically, it generates `assert(false)` followed by `assume(false)`, I think this piece of detail is worth mentioning since it may cause some asserts to become unreachable, and hence vacuously pass.

> statement when compiling any unsupported feature. This will cause proofs to fail

Reviewer 2: ```suggestion
statement when compiling any unsupported feature. This will cause verification to fail
```

if the statement is reachable during the verification stage. However, the

> analysis will not affected if the statement is not reachable from the code under

Reviewer 3: analysis will not *be* affected

> verification, so users can still verify components of their code not using

Reviewer 3: ~not using~ that do not use

(I think your phrasing is technically fine, it is just potentially ambiguous what 'not using' refers to (the code, not the verification) until you read after it in the sentence.)

(Hopefully not too nitpicky here, but it's public facing documentation, so I'm looking for these things)

unsupported features.

> ### Assembly

Reviewer 2: I think the official term is "Inline Assembly".

RMC does not support assembly code for now. We may add it in the future but at
present there are no plans to do so.

Check out the tracking issues for [inline assembly (`asm!`
macro)](https://github.com/model-checking/rmc/issues/2) and [global assembly
(`asm_global!` macro)](https://github.com/model-checking/rmc/issues/316) to know
more about the current status.

### Concurrency

> Concurrent features are out of scope for RMC. In general, the verification of

Reviewer 2: Should we add "currently"?
```suggestion
Concurrent features are currently out of scope for RMC. In general, the verification of
```

concurrent programs continues to be an open research problem where most tools
that analyze concurrent code lack support for other features. Because of this,
RMC emits a warning whenever it encounters concurrent code and compiles as if it

> was sequential code.

Reviewer 2: Does RMC pick some order for execution of concurrent code? If so, we should point out that this may lead to RMC missing bugs that may occur with a different execution order.

Author: AFAIK it assumes all code to be sequential. I think there are some CBMC options for concurrent code verification but we do not make use of them at the moment.

### Standard library functions

At present, RMC is able to link in functions from the standard library but the

> generated code will not contain them. This results in verification failures if

Reviewer 2: > but the generated code will not contain them

This sentence a bit vague. Can we clarify it? Does RMC generate these functions with an empty body?

Author: Added a sentence to mention these are treated in a similar way to unsupported features. Let me know what you think!

the code under verification, for example, includes a `println!` statement.

We have done some experiments to embed the standard library into the generated
code, but this causes verification times to increase significantly. As of now,
we have not been able to find a simple solution for [this
issue](https://github.com/model-checking/rmc/issues/581), but we have some ideas
for future work in this direction.

> ### Advanced features

Reviewer 3: I worry this section might be too vague. It might be a good idea to back-reference what we say in the table? (E.g. there's a lot of links from the table to here that seem to be about projections, and I think our goal is to have to solid by 1.0.)

I think over-technical would be better than vague. If this section answered for us "what would it take for us to remove this section from this doc because we know we support it now" I think that's be better than being vague.

Author: I was not sure how to make this more technical, so I highlighted some outstanding issues here.

Please let me know how it looks and if it does not look good to you we can work together on this section.

unaligned_volatile_store | No | See [Observations - Concurrency](#concurrency) |
unchecked_add | Yes | |
unchecked_div | Yes | |
unchecked_mul | Yes | |
unchecked_rem | Yes | |
unchecked_shl | Yes | |
unchecked_shr | Yes | |
unchecked_sub | Yes | |
unlikely | Yes | |
unreachable | Yes | |
variant_count | No | |
volatile_copy_memory | Partial | See [Observations - Concurrency](#concurrency) |
volatile_copy_nonoverlapping_memory | Partial | See [Observations - Concurrency](#concurrency) |
volatile_load | Partial | See [Observations - Concurrency](#concurrency) |
volatile_set_memory | No | See [Observations - Concurrency](#concurrency) |
volatile_store | No | See [Observations - Concurrency](#concurrency) |
wrapping_add | Yes | |
wrapping_mul | Yes | |
wrapping_sub | Yes | |
write_bytes | Yes | |

#### Atomics

All atomic intrinsics are compiled as an atomic block where the operation is
performed. But as noted in [Observations - Concurrency](#concurrency), CBMC
support for concurrent verification is limited and not used by default.
Verification on code containing atomic intrinsics should not be trusted given
that CBMC assumes the code to be sequential.

#### SIMD instructions

> While RMC is capable of generating code for SIMD instructions. Unfortunately,

Reviewer 2: ```suggestion
While RMC is capable of generating code for SIMD instructions, unfortunately,
```

CBMC does not provide support for some operations like vector comparison (e.g.,

> `simd_eq`) or shuffling (e.g., `simd_shuffle8`).

Reviewer 3: wait, I thought I added shuffle.

Author: Sorry, I thought this was not supported by CBMC but after reviewing related issues I see that you implemented shuffle operations in RMC. So I removed shuffling from here.

statement is reachable during the verification stage, while `assume(false)` will
block any further exploration of the path. However, the analysis will not be
affected if the statement is not reachable from the code under verification, so
users can still verify components of their code that do not use unsupported
features.

### Assembly

RMC does not support assembly code for now. We may add it in the future but at
present there are no plans to do so.

Check out the tracking issues for [inline assembly (`asm!`
macro)](https://github.com/model-checking/rmc/issues/2) and [global assembly
(`asm_global!` macro)](https://github.com/model-checking/rmc/issues/316) to know
more about the current status.

### Concurrency

Concurrent features are currently out of scope for RMC. In general, the
verification of concurrent programs continues to be an open research problem
where most tools that analyze concurrent code lack support for other features.
Because of this, RMC emits a warning whenever it encounters concurrent code and
compiles as if it was sequential code.

### Standard library functions

At present, RMC is able to link in functions from the standard library but the

> generated code will not contain them unless they generic, intrinsics, inlined or

Reviewer 2: ```suggestion
generated code will not contain them unless they are generic, intrinsics, inlined or
```

Author: Thanks!

sinf32 | Yes | |
sinf64 | Yes | |
size_of | Yes | |
size_of_val | Yes | |
sqrtf32 | Yes | |
sqrtf64 | Yes | |
sub_with_overflow | Yes | |
transmute | Yes | |
truncf32 | Yes | |
truncf64 | Yes | |
try | No | |
type_id | Yes | |
type_name | Yes | |
unaligned_volatile_load | Partial | See [Notes - Concurrency](#concurrency) |
unaligned_volatile_store | No | See [Notes - Concurrency](#concurrency) |
unchecked_add | Yes | |
unchecked_div | Yes | |
unchecked_mul | Yes | |
unchecked_rem | Yes | |
unchecked_shl | Yes | |
unchecked_shr | Yes | |
unchecked_sub | Yes | |
unlikely | Yes | |
unreachable | Yes | |
variant_count | No | |
volatile_copy_memory | Partial | See [Notes - Concurrency](#concurrency) |
volatile_copy_nonoverlapping_memory | Partial | See [Notes - Concurrency](#concurrency) |
volatile_load | Partial | See [Notes - Concurrency](#concurrency) |
volatile_set_memory | No | See [Notes - Concurrency](#concurrency) |
volatile_store | No | See [Notes - Concurrency](#concurrency) |
wrapping_add | Yes | |
wrapping_mul | Yes | |
wrapping_sub | Yes | |
write_bytes | Yes | |

#### Atomics

All atomic intrinsics are compiled as an atomic block where the operation is

> performed. But as noted in [Notes - Concurrency](#concurrency), CBMC

Reviewer 2: Should we replace the two occurrences of CBMC in this section by RMC?

Author: Yes, done.
